### PR TITLE
Adding GreatExpectationsBigQueryOperator

### PR DESCRIPTION
**Pull request by brian-lavery**
This is my second attempt to submit a request to add the GreatExpectationsBigQueryOperator to Airflow. There were some problems with my first attempt.  

I'm probably bending protocol a bit because this is what I call a 'preliminary' pull request. Before going to the effort of writing the tests for the new operator and submitting a more refined pull request, I need feedback about the viability of this operator being accepted into Airflow and any other major flaws I need to address.

I'll start with some background. What is Great Expectations(GE)? It's an open source python project that automates data QA testing and data profiling. A user's data '[expectations](https://docs.greatexpectations.io/en/latest/reference/glossary_of_expectations.html)' can be laid out in a json file and GE will check those expectations against a table or query result by firing off a series of SQL queries. Results of the checks are output in json and html files. I've created a [short video](https://photos.app.goo.gl/PPHgCySA6iprVSWz5) to explain how GE and the operator work.

I'm looking for any feedback I can get but I have three immediate questions:
1. Does this sound like a contribution that would be accepted into Airflow?
2. Is there any special testing that needs to be done because the operator imports parts of the great-expectations package (version 0.12.1), which has never been used before in Airflow?
3. Is it a problem that the operator connects to BigQuery in a somewhat unusual way compared with other operators? The operator takes a BigQuery connection as a parameter but the actual connection to BigQuery is made inside of the GE code. Therefore, the operator pulls some connection info from the hook but just passes that information to GE so it can make the connection.

Thanks,
Brian

### CODE REVIEW

File: airflow/providers/google/cloud/operators/bigquery.py [hunk 1]
```diff
@@ -2109,3 +2128,265 @@ def execute(self, context: Any):
     def on_kill(self):
         if self.job_id and self.cancel_on_kill:
             self.hook.cancel_job(job_id=self.job_id, project_id=self.project_id, location=self.location)
+
+
+class GreatExpectationsBigQueryOperator(BaseOperator):
+    """
+         Use Great Expectations to validate data expectations against a BigQuery table or the result of a SQL query.
```
Comment (turbaszek) [inline airflow/providers/google/cloud/operators/bigquery.py]:
```suggestion
    """
    Use Great Expectations to validate data expectations against a BigQuery table or the result of a SQL query.
```
We should avoid redundant indentation

File: airflow/providers/google/cloud/operators/bigquery.py [hunk 2]
```diff
@@ -2109,3 +2128,265 @@ def execute(self, context: Any):
...
+        :type fail_if_expectations_not_met: boolean
+    """
+
+    @apply_defaults
+    def __init__(self, gcp_project, expectations_file_name, gcs_bucket, gcs_expectations_prefix,
+                 gcs_validations_prefix, gcs_datadocs_prefix, validation_type, validation_type_input,
+                 bq_dataset_name, email_to, send_alert_email=True, include_datadocs_link_in_email=False,
+                 fail_if_expectations_not_met=True,
+                 bigquery_conn_id='bigquery_default',
+                 *args, **kwargs):
```
Comment (turbaszek) [inline airflow/providers/google/cloud/operators/bigquery.py]: Please apply `pre-commit run black` to this file. We started using black formatter for providers package. Also please add type hints and as only keywords are allowed in operators we should remove `*args` and use the following signature `__init__(self, *, gcp_project, ..., **kwargs)` In this way python will know that only keywords are allowed (take a look at other operators).

File: airflow/providers/google/cloud/operators/bigquery.py [hunk 3]
```diff
@@ -2109,3 +2128,265 @@ def execute(self, context: Any):
...
+        self.expectations_file_name = expectations_file_name
+        if validation_type.upper() not in GreatExpectationsValidations:
+            raise AirflowException("argument 'validation_type' must be one of %r." % great_expectations_valid_type)
```
Comment (turbaszek) [inline airflow/providers/google/cloud/operators/bigquery.py]: We can use f-strings now

File: airflow/providers/google/cloud/operators/bigquery.py [hunk 4]
```diff
@@ -2109,3 +2128,265 @@ def execute(self, context: Any):
...
+        self.gcs_bucket = gcs_bucket
+        self.gcs_expectations_prefix = gcs_expectations_prefix
+        self.gcs_validations_prefix = gcs_validations_prefix
+        self.gcs_datadocs_prefix = gcs_datadocs_prefix
+        self.email_to = email_to
+        self.send_alert_email = send_alert_email
+        self.include_datadocs_link_in_email = include_datadocs_link_in_email
+        self.fail_if_expectations_not_met = fail_if_expectations_not_met
+
+        super(GreatExpectationsBigQueryOperator, self).__init__(*args, **kwargs)
```
Comment (turbaszek) [inline airflow/providers/google/cloud/operators/bigquery.py]:
```suggestion
        super(self).__init__(*args, **kwargs)
```

File: airflow/providers/google/cloud/operators/bigquery.py [hunk 5]
```diff
@@ -2109,3 +2128,265 @@ def execute(self, context: Any):
...
+
+    def execute(self, context):
+        conn = BaseHook.get_connection(self.bigquery_conn_id)
+        connection_json = json.loads(conn.extra)
```
Comment (turbaszek) [inline airflow/providers/google/cloud/operators/bigquery.py]: There's a method for that:
```suggestion
        connection_json = conn.extra_dejson
```

File: airflow/providers/google/cloud/operators/bigquery.py [hunk 6]
```diff
@@ -2109,3 +2128,265 @@ def execute(self, context: Any):
...
+                        'prefix': self.gcs_datadocs_prefix
+                    },
+                    "site_index_builder": {
+                        "class_name": "DefaultSiteIndexBuilder",
+                    },
+                }
+            },
+            config_variables_file_path=None,
+            commented_map=None,
+        )
```
Comment (turbaszek) [inline airflow/providers/google/cloud/operators/bigquery.py]: Do I correctly understand that those are configuration for Great Expectations? If yes, do they allow users to pass all possible configuration arguments? Currently, we prefer (at least in GCP ops) interfaces that requires users to pass dictionaries instead of individual keys.

File: airflow/providers/google/cloud/operators/bigquery.py [hunk 7]
```diff
@@ -2109,3 +2128,265 @@ def execute(self, context: Any):
...
+            assets_to_validate=[batch],
+            run_id=run_id)
+
+        validation_result_identifier = list(results['run_results'].keys())[0]
+        # For the given validation_result_identifier, get a link to the data docs that were generated by Great
+        # Expectations as part of the validation.
+        data_docs_url = \
+            data_context.get_docs_sites_urls(resource_identifier=validation_result_identifier, site_name='GCS_site')[0][
+                'site_url']
+        log.info("Data docs url is: " + data_docs_url)
```
Comment (turbaszek) [inline airflow/providers/google/cloud/operators/bigquery.py]:
```suggestion
        log.info("Data docs url is: %s", data_docs_url)
```
To be pylint compatible

File: airflow/providers/google/cloud/operators/bigquery.py [hunk 8]
```diff
@@ -2109,3 +2128,265 @@ def execute(self, context: Any):
...
+                self.send_alert(data_docs_url)
+            if self.fail_if_expectations_not_met:
+                raise AirflowException('One or more expectations were not met')
+
+    # Generate a unique name for a temporary BQ table
+    def get_temp_table_name(self):
+        now = datetime.datetime.now().strftime('%Y%m%d_%H%M%S')
+        name_start = "temp_great_expectations_" + now + '_'
+        full_name = name_start + ''.join(random.choices(string.ascii_uppercase +
+                                                        string.digits, k=5))
```
Comment (turbaszek) [inline airflow/providers/google/cloud/operators/bigquery.py]: How about using uuid? We can shorten it if necessary.

File: airflow/providers/google/cloud/operators/bigquery.py [hunk 9]
```diff
@@ -2109,3 +2128,265 @@ def execute(self, context: Any):
...
+        log.info("Generated name for temporary table: " + full_name)
```
Comment (turbaszek) [inline airflow/providers/google/cloud/operators/bigquery.py]:
```suggestion
        log.info("Generated name for temporary table: %s", full_name)
```

File: airflow/providers/google/cloud/operators/bigquery.py [hunk 10]
```diff
@@ -2109,3 +2128,265 @@ def execute(self, context: Any):
...
+        return full_name
+
+    def send_alert(self, data_docs_url):
+        if self.include_datadocs_link_in_email:
+            # From an Airflow variable set by the user, get the domain name of the service serving the data docs.
+            domain = Variable.get("great_expectations_datadocs_domain")
```
Comment (turbaszek) [inline airflow/providers/google/cloud/operators/bigquery.py]: This value should be configurable on operator level

File: airflow/providers/google/cloud/operators/bigquery.py [hunk 11]
```diff
@@ -2109,3 +2128,265 @@ def execute(self, context: Any):
...
+                                Great Expectations Alert<br>
+                                One or more data expectations were not met in the {0} file. {1}
+                           </div>
+                  </body>
+                </html>
+                '''.format(self.expectations_file_name, results)
+        send_email(self.email_to, 'expectations in ' + self.expectations_file_name + ' not met', email_content,
+                   files=None, cc=None, bcc=None,
+                   mime_subtype='mixed', mime_charset='us_ascii')
+
```
Comment (turbaszek) [inline airflow/providers/google/cloud/operators/bigquery.py]: How about extracting this string to a top level as a constant? This should improve readability of operator logic

File: airflow/providers/google/cloud/operators/bigquery.py [hunk 12]
```diff
@@ -20,33 +20,52 @@
...
 import uuid
 import warnings
 from typing import Any, Dict, Iterable, List, Optional, Sequence, Set, SupportsAbs, Union
+from urllib.parse import urlsplit
 
 import attr
 from google.api_core.exceptions import Conflict
 from google.cloud.bigquery import TableReference
 
+from great_expectations.data_context.types.base import DataContextConfig
```
Comment (dlamblin) [inline airflow/providers/google/cloud/operators/bigquery.py]: The dependency will need to get added as an extra to setup.py

Comment (turbaszek) [reply airflow/providers/google/cloud/operators/bigquery.py]: I'm wondering how we should approach it. GE should not be a requirement for other Google operators, on the other hand having separate package for GE doesn't sounds like a good way. Unless we make this operator generic and provider neutral

File: airflow/providers/google/cloud/operators/bigquery.py [hunk 13]
```diff
@@ -2109,3 +2128,265 @@ def execute(self, context: Any):
     def on_kill(self):
         if self.job_id and self.cancel_on_kill:
             self.hook.cancel_job(job_id=self.job_id, project_id=self.project_id, location=self.location)
+
+
+class GreatExpectationsBigQueryOperator(BaseOperator):
```
Comment (dlamblin) [inline airflow/providers/google/cloud/operators/bigquery.py]: I'm not much of a committer so take this with a grain of salt as a user suggestion, but:
I'd like to see this broken into two classes, `GreatExpectationsBaseOperator`, and `GreatExpectationsBigQueryOperator`.

The former could have a method `_configure_project(…)` composed of calls to `_configure_datasouces(…)` `_configure_stores(…)` and `_configure_docs(…)` which would effectively throw a not implemented exception. Its init would not mention the gcp specific params but still cover the validations, email, and other parameters that are general. The latter could then implement these methods to build the config used in the base's execute method. Its init could call super with kwargs. This would leave it open to users if they wanted to run expectations on one of the other supported datasources to implement something that could.

I'm a little unsure of how you could support EG a user that wants to DQ check a mysql db and have the stores and docs in gcp vs checking a mysql db with the stores and docs on aws. I almost want to recommend mixins, but am unsure.

File: airflow/providers/google/cloud/operators/bigquery.py (hunk 6 above)
Comment (feluelle) [reply airflow/providers/google/cloud/operators/bigquery.py]: Yes Tomek, I already played around with GE for some time but never directly used the `DataContextConfig`.

This might be a good option to use instead of having to build a whole file structure through `great_expectations init` and having to manage those yaml files. I can see that work, but I would give the caller of this operator more control about the sections like datasources, validation_operators, stores and data_docs_sites.

I've already read @dlamblin comment on adding more "abstraction layers" when we want to introduce GE to the project. I am all for it. I think this would make the integration in for example BigQuery a lot easier if we have such a base GE operator.

File: airflow/providers/google/cloud/operators/bigquery.py (hunk 13 above)
Comment (feluelle) [reply airflow/providers/google/cloud/operators/bigquery.py]: All for it.

Comment (turbaszek) [reply airflow/providers/google/cloud/operators/bigquery.py]: So this, sounds like an answer to https://github.com/apache/airflow/pull/11113#discussion_r494165469 - I would be in favor of having generic GE. I'm not familiar with this tool but from https://github.com/apache/airflow/pull/11113#discussion_r494205186 it look like we should be able to create one

Comment (brian-lavery) [reply airflow/providers/google/cloud/operators/bigquery.py]: Ok, so we'll work on creating a `GreatExpectationsBaseOperator` and extending it using the `GreatExpectationsBigQueryOperator`.

**Comment by ryw** [comment]
@mrrobby could you jump in on this PR?

File: airflow/providers/google/cloud/operators/bigquery.py [hunk 14]
```diff
@@ -2109,3 +2122,287 @@ def execute(self, context: Any):
...
+        :param datadocs_domain: The domain from which the data docs are set up to be served (e.g. ge-data-docs-dot-my-gcp-project.ue.r.appspot.com).
+            This only needs to be set if datadocs_link_in_email is set to True.
+        :type datadocs_domain: str
+        :param email_to:  Email address to receive any alerts when expectations are not met.
+        :type email_to: str
+        :param fail_if_expectations_not_met: Fail the Airflow task if expectations are not met?  Defaults to True.
+        :type fail_if_expectations_not_met: boolean
+    """
+
+    _EMAIL_CONTENT = '''
```
Comment (brian-lavery) [inline airflow/providers/google/cloud/operators/bigquery.py]: @turbaszek , as suggested, the email content was made a constant

File: airflow/providers/google/cloud/operators/bigquery.py [hunk 15]
```diff
@@ -2109,3 +2122,287 @@ def execute(self, context: Any):
...
+                          <div style="margin-left: 35px;">
+                            Great Expectations Alert<br>
+                            One or more data expectations were not met in the {0} file. {1}
+                       </div>
+              </body>
+            </html>
+            '''
+
+    @apply_defaults
+    def __init__(self, *, gcp_project, expectations_file_name, gcs_bucket, gcs_expectations_prefix,
```
Comment (brian-lavery) [inline airflow/providers/google/cloud/operators/bigquery.py]: @turbaszek , as you can see, I did not yet format the file but I will when I check in changes with the tests.

File: airflow/providers/greatexpectations/operators/greatexpectations_base.py [hunk 16]
```diff
@@ -0,0 +1,30 @@
+from airflow.models import BaseOperator
+
+
+class GreatExpectationsBaseOperator(BaseOperator):
```
Comment (brian-lavery) [inline airflow/providers/greatexpectations/operators/greatexpectations_base.py:31]: @turbaszek , this is the new GreatExpectationsBaseOperator.  It's really just an interface right now.  That was the approach recommended by the GE team because they are planning to do a big refactor and it's not worth trying to move any generic functionality into the base operator until the refactor is done.

File: airflow/providers/google/cloud/operators/bigquery.py [hunk 17]
```diff
@@ -2109,3 +2122,287 @@ def execute(self, context: Any):
...
+                 fail_if_expectations_not_met=True,
+                 bigquery_conn_id='bigquery_default', **kwargs):
+
+        super().__init__(**kwargs)
+
+        great_expectations_valid_type = set(item.value for item in GreatExpectationsValidations)
+
+        self.expectations_file_name = expectations_file_name
+        if validation_type.upper() not in GreatExpectationsValidations:
+            raise AirflowException(f"argument 'validation_type' must be one of {great_expectations_valid_type}")
```
Comment (brian-lavery) [inline airflow/providers/google/cloud/operators/bigquery.py]: @turbaszek , as suggested, changed this to use f-strings.

File: airflow/providers/google/cloud/operators/bigquery.py [hunk 18]
```diff
@@ -2109,3 +2122,287 @@ def execute(self, context: Any):
...
+                            One or more data expectations were not met in the {0} file. {1}
+                       </div>
+              </body>
+            </html>
+            '''
+
+    @apply_defaults
+    def __init__(self, *, gcp_project, expectations_file_name, gcs_bucket, gcs_expectations_prefix,
+                 gcs_validations_prefix, gcs_datadocs_prefix, validation_type, validation_type_input,
+                 bq_dataset_name, email_to, datadocs_domain='none', send_alert_email=True, datadocs_link_in_email=False,
```
Comment (brian-lavery) [inline airflow/providers/google/cloud/operators/bigquery.py]: @turbaszek , as suggested, made the datadocs domain a parameter instead of setting it from an Airflow variable.

File: airflow/providers/google/cloud/operators/bigquery.py [hunk 19]
```diff
@@ -2109,3 +2122,287 @@ def execute(self, context: Any):
...
+                self.log.info('Sending alert email...')
+                self.send_alert(data_docs_url)
+            if self.fail_if_expectations_not_met:
+                raise AirflowException('One or more expectations were not met')
+
+    # Generate a unique name for a temporary BQ table
+    def get_temp_table_name(self):
+        now = datetime.datetime.now().strftime('%Y%m%d_%H%M%S')
+        name_start = "temp_great_expectations_" + now + '_'
+        full_name = name_start + ''.join(str(uuid.uuid4().hex))
```
Comment (brian-lavery) [inline airflow/providers/google/cloud/operators/bigquery.py]: @turbaszek , as suggested, changed the temp table name to be generated based on a uuid.

File: airflow/providers/google/cloud/operators/bigquery.py [hunk 20]
```diff
@@ -2109,3 +2122,287 @@ def execute(self, context: Any):
...
+                        },
+                        {
+                            "name": "update_data_docs",
+                            "action": {"class_name": "UpdateDataDocsAction"},
+                        },
+                    ],
+                }
+            },
+            stores={
+                'expectations_GCS_store': {
```
Comment (mrrobby) [inline airflow/providers/google/cloud/operators/bigquery.py]: I feel like setting the expectations store and the validations store (or all of the data context at that) should be abstracted away from here somehow, perhaps by creating a great expectations hook. That would allow to, say, use a gcs bucket for these stores and create a snowflake great expectations operator perhaps.
```python
datasource = ge_gcs_hook.create_bigquery_datasource(bigquery_conn_id, bq_dataset_name, other_params) 
data_context = ge_gcs_hook.create_data_context(datasource, gcs_conn_id, gcs_bucket, gcs_validations_prefix, gcs_expectations_prefix)
```
so we can also have something like
```python
# Base hook sets abstract on data context
custom_datasource = ge_hook.create_custom_datasource(conn_id, dataset_name, class_name, module_name) 
```
Something like that

File: airflow/providers/google/cloud/operators/bigquery.py [hunk 21]
```diff
@@ -20,26 +20,39 @@
...
 from typing import Any, Dict, Iterable, List, Optional, Sequence, Set, SupportsAbs, Union
+from urllib.parse import urlsplit
 
 import attr
 from google.api_core.exceptions import Conflict
 from google.cloud.bigquery import TableReference
 
+from airflow.providers.greatexpectations.operators.greatexpectations_base import GreatExpectationsBaseOperator
+from great_expectations.data_context.types.base import DataContextConfig
+from great_expectations.data_context import BaseDataContext
```
Comment (turbaszek) [inline airflow/providers/google/cloud/operators/bigquery.py]: I'm not sure about it. Having GE should not be required for BigQuery operators. I think we can treat GE as separate provider and keep everything related to it there - including BigQuery specific operators. WDYT @mik-laj @kaxil @feluelle ?

File: airflow/providers/google/cloud/operators/bigquery.py [hunk 22]
```diff
@@ -2109,3 +2122,287 @@ def execute(self, context: Any):
...
+
+        if self.validation_type == GreatExpectationsValidations.SQL.value:
+            batch_kwargs["query"] = self.validation_type_input
+            batch_kwargs["data_asset_name"] = self.bq_dataset_name
+            batch_kwargs["bigquery_temp_table"] = self.get_temp_table_name()
+        elif self.validation_type == GreatExpectationsValidations.TABLE.value:
+            batch_kwargs["table"] = self.validation_type_input
+            batch_kwargs["data_asset_name"] = self.bq_dataset_name
+
+        log.info("batch_kwargs: " + str(batch_kwargs))
```
Comment (turbaszek) [inline airflow/providers/google/cloud/operators/bigquery.py]:
```suggestion
        self.log.info("batch_kwargs: " + str(batch_kwargs))
```
This should do and probably there will be no need to have `log = logging.getLogger()` in top level

File: airflow/providers/greatexpectations/operators/greatexpectations_base.py (hunk 16 above)
Comment (turbaszek) [reply airflow/providers/greatexpectations/operators/greatexpectations_base.py:31]: Then I would say there's no need for such an operator - it gives us nothing currently. We should add the operator / hook only if it brings some value, for example creates repository of reusable methods. In this case, this may only make future refactoring harder.

Comment (spbail) [reply airflow/providers/greatexpectations/operators/greatexpectations_base.py:31]: Just to chime in, one challenge we're looking at is that GE can basically work with any kind of backend, and we have a lot of folks who've built their own Airflow operators already. Even if the base operator isn't super helpful as part of this PR, I can see us extending it to work with other backends very quickly.

I think this would be my biggest question for @ryw and @mrrobby. I'm thinking the base operator could implement a generic load data context + load batch + validation flow similar to a draft PR we received: https://github.com/apache/airflow/pull/11113/files

Comment (turbaszek) [reply airflow/providers/greatexpectations/operators/greatexpectations_base.py:31]: > the base operator could implement a generic load data context + load batch + validation

I agree with that, the base operator should implement methods that can be reused in other more specific operators. It should also implement some error handling as well as auth part if necessary

File: airflow/providers/google/cloud/operators/bigquery.py (hunk 21 above)
Comment (feluelle) [reply airflow/providers/google/cloud/operators/bigquery.py]: +1 for moving all `great_expectations` imports to its provider package.

File: airflow/providers/google/cloud/operators/bigquery.py [hunk 23]
```diff
@@ -2109,3 +2122,287 @@ def execute(self, context: Any):
...
+         are output to GCS in both JSON and HTML formats.
+         Here's the current list of expectations types:
+         https://docs.greatexpectations.io/en/latest/reference/glossary_of_expectations.html
+         Here's how to create expectations files:
+         https://docs.greatexpectations.io/en/latest/guides/tutorials/how_to_create_expectations.html
+        :param gcp_project:  The GCP project which houses the GCS buckets where the expectations files are stored
+            and where the validation files & data docs will be output (e.g. HTML docs showing if the data matches
+            expectations).
+        :type gcp_project: str
+        :param expectations_file_name: The name of the JSON file containing the expectations for the data.
```
Comment (spbail) [inline airflow/providers/google/cloud/operators/bigquery.py]: Just noticed this here, is there a specific reason for using the filename rather than the expectation suite name? I think that would be more in line with how we've been referring to expectation suites, and you should be able to load the expectation suite from the context by name.

File: airflow/providers/google/cloud/operators/bigquery.py [hunk 24]
```diff
@@ -98,32 +98,26 @@ class BigQueryCheckOperator(CheckOperator):
     a sql query that will return a single row. Each value on that
     first row is evaluated using python ``bool`` casting. If any of the
     values return ``False`` the check is failed and errors out.
-
```
Comment (feluelle) [inline airflow/providers/google/cloud/operators/bigquery.py:101]: Hey @brian-lavery can you also revert the changes made to this file, please?

**Comment by spbail** [comment]
Hi all, I just wanted to give a quick update! After some conversations internally with the Great Expectations team, we figured it would make the most sense for us to submit the base operator that actually has quite solid functionality. Any more specific operators like this one here can then inherit from it. I discussed that with Brian and he's on board. Apologies for the back and forth, this has been going through a few revisions and we've finally settled on a path.

**Comment by brian-lavery** [comment]
@turbaszek and @feluelle , given that @spbail and the GE team will be submitting a new base operator for GE, I'm going to close this pull request.  I'll open a new one for the GreatExpectationsBigQueryOperator after Sam submits her pull request and extend her base operator to create the GreatExpectationsBigQueryOperator.
